### PR TITLE
Bug 1218126 - login needs to make use of token if provided

### PR DESCRIPTION
**Pull request by fabianofranz**

### CODE REVIEW

**Comment by fabianofranz** [comment]
@deads2k Please take a look, this fixes https://bugzilla.redhat.com/show_bug.cgi?id=1218126. I wasn't sure about exiting `gatherAuthInfo` if the provided token is good ([this version](https://github.com/openshift/origin/pull/2257/files#diff-f260fb37cbe8b4d20167cc6bc6d4e51aR158)) or jump into the check for matching user stanzas logic. 

Notice we can end with multiple user stanzas with the same token, but that's a different issue (related to https://github.com/openshift/origin/issues/2142) - also happens in multiple `osc login` with interactive username and pass provided.

**Comment by fabianofranz** [comment]
@liggit fyi

**Comment by fabianofranz** [comment]
@liggitt fyi

**Comment by liggitt** [comment]
can we add a test to make sure this doesn't break again?

**Comment by liggitt** [comment]
@deads2k has a pull in progress (https://github.com/openshift/origin/pull/2226) to pin user stanza names based on the username returned by the server... that should fix the issue with multiple stanzas for the same token

**Comment by liggitt** [comment]
I thought this was working before because `--token` got picked up as an override for the kubeconfig processing... are overrides to that no longer working for `login`?

**Comment by fabianofranz** [comment]
@liggitt I think the `--token` override is there, but it mostly fails because before trying a `whoAmI` with the provided token we do [this](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/loginoptions.go#L191).

**Comment by fabianofranz** [comment]
In fact maybe we don't even need to logic around storing the token in `LoginOptions`, just try a `whoAmI` before anything else in `gatherAuthInfo` and if it succeeds set the `Name` and return. I'll test that.

**Comment by liggitt** [comment]
> ... just try a whoAmI before anything else in gatherAuthInfo and if it succeeds set the Name and return

need to be careful about that... what would you expect to happen if you did `osc login` with no args and you were already logged in? no-op with message "You are already logged in"? If so, how do you login as someone else?

**Comment by liggitt** [comment]
I like to let #2226 merge, then take a look at this

**Comment by liggitt** [comment]
looking at the code, I have no idea how `osc login --token=...` worked, but I promise it did... it would log you in as the user the token was valid for and tell you "Already logged into X as Y"

**Comment by fabianofranz** [comment]
@liggitt Yep, I remember it working at least in the first versions of `login`, got broken at some point in the middle by a refactoring or something. Meaning we need tests.  :)

**Comment by fabianofranz** [comment]
Sounds good, I'll wait for https://github.com/openshift/origin/pull/2226 and check how it looks after that.

**Comment by fabianofranz** [comment]
@liggitt @deads2k please review.

File: pkg/cmd/cli/cmd/login.go [hunk 1]
```diff
@@ -169,6 +171,10 @@ func (o *LoginOptions) Complete(f *osclientcmd.Factory, cmd *cobra.Command, args
 	o.CAFile = kcmdutil.GetFlagString(cmd, "certificate-authority")
 	o.InsecureTLS = kcmdutil.GetFlagBool(cmd, "insecure-skip-tls-verify")
 
+	if token := kcmdutil.GetFlagString(cmd, "token"); len(token) > 0 {
```
Comment (deads2k) [inline pkg/cmd/cli/cmd/login.go]: I think you can assign this directly.

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 2]
```diff
@@ -81,7 +81,10 @@ func (o *LoginOptions) getClientConfig() (*kclient.Config, error) {
 				}
 			}
 		}
+	}
 
+	if len(o.Token) > 0 {
```
Comment (deads2k) [inline pkg/cmd/cli/cmd/loginoptions.go]: Move to `getAuthInfo`

Comment (deads2k) [inline pkg/cmd/cli/cmd/loginoptions.go]: And only assign it to our stateful `clientConfig` if it works.

**Comment by deads2k** [comment]
Minor comments.  After that, lgtm.

**Comment by fabianofranz** [comment]
Fixed comments, [merge].

**Comment by fabianofranz** [comment]
re[merge]

**Pull request merged**